### PR TITLE
kubeadm: Add options to specify container registry username and password

### PR DESCRIPTION
**Pull request by thomasmey**
This enables private container registries to be used, which are protceted.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by neolit123 — COMMENTED** [review]
hi, i don't think i'm in favor of this change given crictl can be configured with a config file and environment variables. please see:
https://github.com/kubernetes/kubeadm/issues/2667
https://github.com/kubernetes-sigs/cri-tools/pull/904

also, kubeadm configuring CRI has been considered out of scope.

File: cmd/kubeadm/app/apis/kubeadm/v1beta3/types.go [hunk 1]
```diff
@@ -128,6 +128,14 @@ type ClusterConfiguration struct {
 	// +optional
 	ImageRepository string `json:"imageRepository,omitempty"`
 
+	// ImageRepositoryUsername username for image repository if any
+	// +optional
+	ImageRepositoryUsername string
+
+	// ImageRepositoryPassword password for image repository if any
+	// +optional
+	ImageRepositoryPassword string
```
Comment (neolit123) [inline cmd/kubeadm/app/apis/kubeadm/v1beta3/types.go:137]: v1beta2 and v1beta3 are locked and new fields cannot be added, thus such a change must be in a future release.
kubeadm APIs are different in this regard to the rest of k8s.

**Comment by thomasmey** [comment]
Hi,

what I want to achieve is: I wand to use kubeadm to pull from private protected repository when I setup new control plane node, but I don't want to persist the credentials on that node.

Is the env AUTH_STRING also passed from kubeadm to crictl? So maybe setting this env before executing kubeadm is what I want, but need to check if I have new enough crictl.

**Comment by thomasmey** [comment]
oh, I just noticed that https://github.com/kubernetes-sigs/cri-tools/pull/904 is very new and not part of any offical release yet, right? Anyway I think this does the trick for me.
I think I did start to work on this actually before https://github.com/kubernetes-sigs/cri-tools/pull/904 was merged, but as I do this in my spare time, so... :-D

**Comment by neolit123** [comment]
persisting secure settings in the cluster configuration has been discussed as out of scope in the past. it is not very secure.

> Is the env AUTH_STRING also passed from kubeadm to crictl? So maybe setting this env before executing kubeadm is what I want, but need to check if I have new enough crictl.

yes, because kubeadm would be the parent process the shells-out to crictl and the ENV variables are inherited.

you could do `AUTH_STRING=foo kubeadm ....` and it should work.

**Comment by neolit123** [comment]
> oh, I just noticed that https://github.com/kubernetes-sigs/cri-tools/pull/904 is very new and not part of any offical release yet, right?

it doesn't seem like the commit is part of a release yet.
https://github.com/kubernetes-sigs/cri-tools/pull/904/commits/511767e596a56badf79c8af022a33e21284c5a1a

you could ask them when a release will happen or build crictl from master.

**Comment by neolit123** [comment]
/close

(triaging some old PRs)

**Pull request closed without merging**